examples/spmd/run_spmd.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import os
+from copy import deepcopy
+from functools import partial
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+from spmd.compiler.log_utils import rank0_debug
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+from spmd import SPMD, Schema
+from spmd.tensor import DeviceMesh, Replicate
+from typing import List, Union, Literal
+
+logger: logging.Logger = logging.getLogger(__name__)
+_debug = partial(rank0_debug, logger)  # type: ignore
+
+# globals --------------
+
+DEVICE_TYPE = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+def setup(rank: int, world_size: int, use_cuda: bool = True) -> None:
+    logging.getLogger().setLevel(
+        logging.DEBUG if rank == 0 else logging.CRITICAL
+    )
+
+    if use_cuda:
+        _debug("--> init process group using nccl")
+        dist.init_process_group("nccl", rank=rank, world_size=world_size)
+        torch.cuda.set_device(rank)
+        print(f"--> device set for rank {rank}")
+    else:
+        _debug("--> init process group using gloo")
+        dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+
+def teardown(rank: int) -> None:
+
+    # Wait for all ranks to reach here before starting shutdown.
+    _debug(f"rank {rank} entering teardown")
+    dist.barrier()
+    dist.destroy_process_group()
+    _debug(f"shut down process group on rank {rank}")
+
+
+def formatted_print(
+    rank: int, name: str, val: int, rank_only: bool = False
+) -> None:
+    if rank_only and not rank == 0:
+        return
+    print(f"{rank} --> {name} = {val}")
+
+
+# --- model
+
+mnist_dims: List[int] = [28 * 28, 500, 250, 100, 50, 25, 10]
+
+
+class MyModel(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.mod = nn.Sequential(
+            *[
+                layer
+                for i in range(2)
+                for layer in [
+                    nn.Linear(mnist_dims[i], mnist_dims[i + 1], bias=True),
+                    nn.ReLU(),
+                ]
+            ]
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.mod(x)
+
+
+class Permute(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.w = torch.nn.Parameter(torch.rand((5, 10)))
+        self.b = torch.nn.Parameter(torch.rand((5)))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_t = x.permute(0, 2, 1)
+        return torch.nn.functional.linear(x_t, self.w, self.b)
+
+
+class replicaModel(nn.Module):
```

Comment:
nit: `ReplicaModel`. Python naming rule.

Reply:
my work in Rust continues to show lol.
fixed and thanks.